### PR TITLE
Plugins: Allow view specific config

### PR DESCRIPTION
**Pull request by nhjschulz**
Add a method for view specfic configuration
keys via json. The goal is to allow additional
settings (i.e. 64x64 layouts) without increasing
the  footprint for others.

A new interface called IJsonConfig defines the API to use. Both plugins and views can implement it. The plugins then forward config calls to the views to have view specific settings covered (if needed).

### CODE REVIEW

**Comment by nhjschulz** [comment]
The change doen't fullfill its goal to save flash right now  👎 
 
Default config:

feature/view_config:
RAM:   [==        ]  18.3% (used 59844 bytes from 327680 bytes)
Flash: [==========]  96.1% (used 1384965 bytes from 1441792 bytes)


Development:
RAM:   [==        ]  18.3% (used 59876 bytes from 327680 bytes)
Flash: [==========]  96.0% (used 1384121 bytes from 1441792 bytes)

File: lib/Plugin/src/PluginWithConfig.hpp [hunk 1]
```diff
@@ -333,6 +318,13 @@ class PluginWithConfig : public Plugin
     bool        m_storeConfigReq;               /**< Is requested to store the configuration in persistent memory? */
     bool        m_reloadConfigReq;              /**< Is requested to reload the configuration from persistent memory? */
 
+    /*TODO REMOVE if all plugins have it. Right now only in DateTimePlugin. */
+    bool mergeConfiguration(JsonObject& jsonMerged, const JsonObjectConst& jsonSource)
```
Comment (BlueAndi) [inline lib/Plugin/src/PluginWithConfig.hpp:322]: What is the idea later with mergeConfiguration() method in this class? Isn't it only used between the plugin main class and its views only?

File: lib/Utilities/src/Util.h [hunk 2]
```diff
@@ -47,6 +47,8 @@
 #include <WString.h>
 #include <cerrno>
 
+#include <YAColor.h>
```
Comment (BlueAndi) [inline lib/Utilities/src/Util.h:50]: library.json dependency update is missing. Now Utilities are depended on YAGfx.

File: lib/Common/src/IJsonConfig.h [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+/******************************************************************************
+ * Types and Classes
+ *****************************************************************************/
+
+/**
+ * This interface defines the functions required for JSON bases configuration
+ * updates. It is used to provide and apply configuration settings to and from
+ * a REST API.
+ */
+class IJsonConfig
```
Comment (BlueAndi) [inline lib/Common/src/IJsonConfig.h:61]: Not sure whether the interface class is necessary, see my comment in PluginWithConfig.hpp
mergeConfiguration() is only required between plugin main class and its view. The other two methods are in general required for configuration file handling in PluginWithConfig.

Additional it increases the dependency from plugin library to common library. But could live in plugin library.

File: lib/Views/src/layouts/DateTimeView64x64.cpp [hunk 4]
```diff
@@ -286,6 +296,74 @@ void DateTimeView64x64::drawAnalogClockHand(YAGfx& gfx, int16_t minute, int16_t
...
+}
+
+bool DateTimeView64x64::mergeConfiguration(JsonObject& jsonMerged, const JsonObjectConst& jsonSource)
+{
+    bool            result          = false;
+    JsonObjectConst jsonAnalogClock = jsonSource["analogClock"];
+
+    if (false == jsonAnalogClock.isNull())
+    {
+        jsonMerged["analogClock"] = jsonAnalogClock;
```
Comment (BlueAndi) [inline lib/Views/src/layouts/DateTimeView64x64.cpp:360]: Did you check the REST API if you change a color value?

File: lib/Views/src/layouts/DateTimeView64x64.cpp [hunk 5]
```diff
@@ -286,6 +296,74 @@ void DateTimeView64x64::drawAnalogClockHand(YAGfx& gfx, int16_t minute, int16_t
...
+
+    for (uint32_t index = 0U;  index < ANA_CLK_COL_MAX; ++index)
+    {
+        jsonAnalogClock[ANALOG_CLOCK_COLOR_KEYS[index]]= Util::colorToHtml(m_analogColors[index]);
+    }
+}
+
+bool DateTimeView64x64::setConfiguration(const JsonObjectConst& jsonCfg)
+{
+    bool result = true;
```
Comment (BlueAndi) [inline lib/Views/src/layouts/DateTimeView64x64.cpp:314]: The result will be true even if "analogClock" would be a string instead an object.

File: lib/Views/src/layouts/DateTimeView64x64.cpp [hunk 6]
```diff
@@ -286,6 +296,74 @@ void DateTimeView64x64::drawAnalogClockHand(YAGfx& gfx, int16_t minute, int16_t
...
+
+bool DateTimeView64x64::mergeConfiguration(JsonObject& jsonMerged, const JsonObjectConst& jsonSource)
+{
+    bool            result          = false;
+    JsonObjectConst jsonAnalogClock = jsonSource["analogClock"];
+
+    if (false == jsonAnalogClock.isNull())
+    {
+        jsonMerged["analogClock"] = jsonAnalogClock;
+        result                    = true;
```
Comment (BlueAndi) [inline lib/Views/src/layouts/DateTimeView64x64.cpp:361]: Be aware, mergeConfiguration() will receive all JSON parameters as String! A conversion might be necessary here.

**Review by BlueAndi — CHANGES_REQUESTED** [review]

**Comment by BlueAndi** [comment]
@nhjschulz Please verify the Swagger API update for DateTimePlugin: https://app.swaggerhub.com/apis/BlueAndi/Pixelix/1.5.0#/DateTimePlugin/getDateTimeCfg

File: lib/Views/src/layouts/DateTimeView64x64.cpp (hunk 4 above)
Comment (nhjschulz) [reply lib/Views/src/layouts/DateTimeView64x64.cpp:360]: Didn't consider REST impact so far, only JSON from web page. Next page has fix for it. It works with this now:

    curl --request POST --url "http://192.168.1.112/rest/api/v1/display/uid/64860/dateTime?analogClock.secondsMode=3&analogClock.handMinCol=%2300000FF"

File: lib/Views/src/layouts/DateTimeView64x64.cpp (hunk 5 above)
Comment (nhjschulz) [reply lib/Views/src/layouts/DateTimeView64x64.cpp:314]: I don't understand this comment. 
Wouldn't RestApiTopicHandler::par2Json() have done this convertion bevore this call already ?

File: lib/Views/src/layouts/DateTimeView64x64.cpp (hunk 6 above)
Comment (nhjschulz) [reply lib/Views/src/layouts/DateTimeView64x64.cpp:361]: I tested  analogClock.secondsMode=3 and analogClock.handMinCol=%2300000FF to work just fine over REST. Also web page still works.

Am I missing something still ?

**Comment by nhjschulz** [comment]
> @nhjschulz Please review the Swagger API update for DateTimePlugin: https://app.swaggerhub.com/apis/BlueAndi/Pixelix/1.5.0#/DateTimePlugin/getDateTimeCfg

grafik

This should be **analogClock.**secondsMode

File: lib/Views/src/layouts/DateTimeView64x64.cpp (hunk 5 above)
Comment (BlueAndi) [reply lib/Views/src/layouts/DateTimeView64x64.cpp:314]: If "analogClock" is Null, the function will return true. But shouldn't it return false in this case?

File: lib/Views/src/layouts/DateTimeView64x64.cpp (hunk 6 above)
Comment (BlueAndi) [reply lib/Views/src/layouts/DateTimeView64x64.cpp]: Ok, then its fine I think.

**Review by BlueAndi — CHANGES_REQUESTED** [review]

File: lib/Views/src/layouts/DateTimeView64x64.cpp (hunk 5 above)
Comment (nhjschulz) [reply lib/Views/src/layouts/DateTimeView64x64.cpp:314]: Will change to fail on missing data. The old code was meant to be "backward compatible" and treated analog clock data as optional.  This makes no sense anymore as this is an analogClock supporting view class now. So the data must be there.

**Review by BlueAndi — APPROVED** [review]

**Pull request merged**